File: functions/src/deliveryPaymentPolicy.ts
/**
 * Server-side mirror of `DeliveryPaymentPolicy`/
 * `InMemoryDeliveryPaymentPolicyRepository` (Dart) — Faz P.1. Same
 * "no Dart<->TypeScript sharing mechanism" duplication precedent as every
 * other mirrored file in this codebase (see `takeawayPricing.ts`'s own
 * doc comment): the Dart file remains the source of truth for the
 * *design* of this policy; this is its server-side enforcement primitive
 * for a future `submitDeliveryOrder` (P.2/P.3, not implemented yet) to
 * consume.
 *
 * **Exports no callable** — this phase deliberately does not expose a
 * production-capable delivery order-submission endpoint (P.1 §6). This
 * module only provides the pure policy check so it doesn't have to be
 * invented later under time pressure.
 *
 * LOCKED (`docs/business_rules.md`, Faz P.1 §2/§3): initial release is
 * COD-only, exactly these 7 method ids — `bank_transfer`/`gift_voucher`/
 * any online-payment method are deliberately excluded. Independent of a
 * `PaymentMethod`'s own `isActive` flag (this module has no notion of
 * that catalog at all) — a caller resolving an actual checkout must
 * additionally check the method's live `isActive` state itself.
 */

export const DELIVERY_ENABLED_PAYMENT_METHOD_IDS: ReadonlySet<string> = new Set([
  "cash",
  "credit_card",
  "pluxee",
  "multinet",
  "setcard",
  "edenred",
  "metropol_card",
]);

export function isEnabledForDeliveryCheckout(paymentMethodId: string): boolean {
  return DELIVERY_ENABLED_PAYMENT_METHOD_IDS.has(paymentMethodId);
}

File: lib/features/payment/domain/models/delivery_payment_policy.dart
import 'payment_method.dart';

/// Which payment methods a customer may choose from at Paket Servis
/// (delivery) checkout — Faz P.1, LOCKED §2/§3: the initial release is
/// COD-only, exactly the 7 methods a
/// [DeliveryPaymentPolicyRepository] seeds by default; online payment,
/// bank transfer, and gift voucher are explicitly excluded.
///
/// **Deliberately independent of [PaymentMethod.isActive]** — that flag
/// controls whether a method exists/works at all, on any channel; this
/// policy controls whether an already-active method is additionally
/// permitted on the delivery channel specifically. A method must satisfy
/// both to actually be offered to a delivery customer — see
/// [isAvailableForDeliveryCheckout]. Never invokes a
/// `PaymentProviderAdapter` — every currently-enabled method is a
/// manually recorded, cash-on-delivery-style instrument.
class DeliveryPaymentPolicy {
  const DeliveryPaymentPolicy({required this.enabledMethodIds});

  /// The [PaymentMethod.id]s currently permitted at delivery checkout.
  final Set<String> enabledMethodIds;

  bool isEnabledForDeliveryCheckout(String paymentMethodId) =>
      enabledMethodIds.contains(paymentMethodId);

  /// The combined check a checkout screen must use — both this policy's
  /// own enablement AND the method's own [PaymentMethod.isActive] must
  /// hold. See this class's own doc comment for why the two are kept
  /// structurally separate rather than collapsed into one flag.
  bool isAvailableForDeliveryCheckout(PaymentMethod method) =>
      method.isActive && isEnabledForDeliveryCheckout(method.id);
}
